### ISSUE TITLE
How to manually reset the model weights of federated averaging algo？

### ISSUE THREAD

**Issue by zzhu2019** [tensorflow/federated#758]
Hi there,

What I want to do now is reset the model using the `model.set_weights()`. But how should I do that, should I recreate both `iterative_process` (by `tff.learning.build_federated_averaging_process(model_fn)`) and `state` (by `iterative_process.initialize()`)? 

UPDATE 6/9/2019:
Well, I just found that there is a method called `tff.utils.update_state()`. Can I use this method to change the model weights without doing anything to the `iterative_process`? Also, for the pass-in parameter `**kwargs`, should I pass key-value pairs into it? And how should I name this key if I wanna change the weights?

UPDATE 7/9/2019:
The `tff.utils.update_state()` seems not to support the AnonymousTuple at the moment. I am a bot confused what the state should be here

**Comment by ZacharyGarrett** [comment]
Hi @zzhu2019, could you help me understand more clearly the context around "reset the module using `model.set_weights()`? If you could share the code you're writing it would help clarify the intent. 

If you want to reset the model weights within the `state` object to the original initialization, calling `iterative_process.initialize()` will produce a new, reset model. Did this not accomplish want you wanted?

**Comment by UltraShieldRog** [comment]
> Hi @zzhu2019, could you help me understand more clearly the context around "reset the module using `model.set_weights()`? If you could share the code you're writing it would help clarify the intent.
> 
> If you want to reset the model weights within the `state` object to the original initialization, calling `iterative_process.initialize()` will produce a new, reset model. Did this not accomplish want you wanted?

Hi Zachary. 

Sorry for the late reply as I cannot log in my account today, I have to reply with my friend's account. Apologize for not being clear, as the `model.set_weights()` is what I found in `tf.keras.layers.Layer.set_weights()` [(tf source code)](https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/keras/engine/base_layer.py#L1090-L1121). To be specific, the reset I mean here is kinda changing the model weights into a set of weights calculated beforehand but not reset to original initialization. The `model.set_weights()` works fine by itself, but when I 

1. initialize a model with `model = tf.keras.models.Sequential([tf.keras.layers.Dense(10, activation=tf.nn.softmax, kernel_initializer='zeros', input_shape=(784,))])`, and 

2. use the `model.set_weights()`, 
3. then `model.compile()`, 
4. after that, use this generated model in `tff.learning.from_compiled_keras_model()` 
5. and then `tff.learning.build_federated_averaging_process()` to produce a new `iterative_process` 
6. and `iterative_process.initialize()` to produce a new state. 
It seems like the model weight is not resetted to the target weights but the initialization weights instead.

Generally, is my understanding of the code correct? And is there a more elegant way to do it?

**Comment by UltraShieldRog** [comment]
> Hi @zzhu2019, could you help me understand more clearly the context around "reset the module using `model.set_weights()`? If you could share the code you're writing it would help clarify the intent.
> 
> If you want to reset the model weights within the `state` object to the original initialization, calling `iterative_process.initialize()` will produce a new, reset model. Did this not accomplish want you wanted?

And, can I use the code below to only update the `state` in order to change the weights without regenerate the `iterative_process`?
`value_dict = {}`
`value_dict['desne/kernal'] = states[edge_id][0]`
`value_dict['desne/bias'] = states[edge_id][1]`
`state = tff.utils.update_state(state=state, **value_dict)`

**Comment by ZacharyGarrett** [comment]
Could you share the code you are using the do steps 1 through 5 in https://github.com/tensorflow/federated/issues/758#issuecomment-530302262? Setting a Keras model's weights with pre-trained weights and then using that in TFF is expected to work (the [Federated Learning for Text Generation tutorial](https://www.tensorflow.org/federated/tutorials/federated_learning_for_text_generation) does this, see the sections on _Load a pre-trained model_ and _Fine-tune the model with Federated Learning_). Lets see what differs from those examples.

**Comment by UltraShieldRog** [comment]
> Could you share the code you are using the do steps 1 through 5 in [#758 (comment)](https://github.com/tensorflow/federated/issues/758#issuecomment-530302262)? Setting a Keras model's weights with pre-trained weights and then using that in TFF is expected to work (the [Federated Learning for Text Generation tutorial](https://www.tensorflow.org/federated/tutorials/federated_learning_for_text_generation) does this, see the sections on _Load a pre-trained model_ and _Fine-tune the model with Federated Learning_). Lets see what differs from those examples.

Basically, I changed a bit about the `model_fn()` as follow:
![image](https://user-images.githubusercontent.com/37321309/64745836-b86ea300-d54c-11e9-8e5d-f472382d731a.png)
And the `model_fn()` is used in:
![image](https://user-images.githubusercontent.com/37321309/64745875-e48a2400-d54c-11e9-8bfc-9add2dd68e72.png)
which produce a new  `iterative_process` and `state`, but the problem is that the state still has the original initialization weights..

**Comment by UltraShieldRog** [comment]
> Could you share the code you are using the do steps 1 through 5 in [#758 (comment)](https://github.com/tensorflow/federated/issues/758#issuecomment-530302262)? Setting a Keras model's weights with pre-trained weights and then using that in TFF is expected to work (the [Federated Learning for Text Generation tutorial](https://www.tensorflow.org/federated/tutorials/federated_learning_for_text_generation) does this, see the sections on _Load a pre-trained model_ and _Fine-tune the model with Federated Learning_). Lets see what differs from those examples.

It seems like there is something wrong with the way I use `model.reset_weights()` cuz the weights inside the state generated by `iterative_process.initialize()` just change correspondingly to the `kernel_initializer` in `model = tf.keras.models.Sequential([
            tf.keras.layers.Dense(
                10, activation=tf.nn.softmax, kernel_initializer='random_uniform', input_shape=(784,))])`

**Comment by UltraShieldRog** [comment]
> Could you share the code you are using the do steps 1 through 5 in [#758 (comment)](https://github.com/tensorflow/federated/issues/758#issuecomment-530302262)? Setting a Keras model's weights with pre-trained weights and then using that in TFF is expected to work (the [Federated Learning for Text Generation tutorial](https://www.tensorflow.org/federated/tutorials/federated_learning_for_text_generation) does this, see the sections on _Load a pre-trained model_ and _Fine-tune the model with Federated Learning_). Lets see what differs from those examples.

Hi Zachary, I am pretty sure that the weight of the model before `model.compile()` inside the `created_compiled_keras_model()` has been changed which can be confirmed by both (`model.weights` and `model.get_weights()`).

**Comment by ZacharyGarrett** [comment]
Its hard to diagnose the issue without a full picture of the situation, the code snippets are helpful but without the complete code I'm still having a hard time fully grasping what the goal is and how it is being attempted.

From what I gather so far this doesn't seem like an issue that needs fixing in TFF, but a question about how to use TFF (or TF) to achieve a particular behavior? If you agree, these kind of questions are better asked at https://stackoverflow.com/tags/tensorflow_federated, which preserves easily searchable database of questions and answers for the community to benefit from.

I'd recommend posting a question there and adding the `keras`, `tensorflow`, and `tensorflow_federated` tags. When posting a question, please try to create a minimal example (from https://github.com/tensorflow/federated/issues/758#issuecomment-531138660 it sounds like only the code inside `def create_compiled_keras_model()` may be necessary?), and post _all_ the code related to that example.

**Comment by UltraShieldRog** [comment]
> Its hard to diagnose the issue without a full picture of the situation, the code snippets are helpful but without the complete code I'm still having a hard time fully grasping what the goal is and how it is being attempted.
> 
> From what I gather so far this doesn't seem like an issue that needs fixing in TFF, but a question about how to use TFF (or TF) to achieve a particular behavior? If you agree, these kind of questions are better asked at https://stackoverflow.com/tags/tensorflow_federated, which preserves easily searchable database of questions and answers for the community to benefit from.
> 
> I'd recommend posting a question there and adding the `keras`, `tensorflow`, and `tensorflow_federated` tags. When posting a question, please try to create a minimal example (from [#758 (comment)](https://github.com/tensorflow/federated/issues/758#issuecomment-531138660) it sounds like only the code inside `def create_compiled_keras_model()` may be necessary?), and post _all_ the code related to that example.

Hi Zachary, thank you for your reply to my question which is not really a bug of so of tff. I have given up changing the model before return the compiled Keras model to generate the new state. But I just found the `tff.learning.state_with_new_model_weights` which can fix my problem in a way more convenient way. I will close this issue soon. : )

**Comment by UltraShieldRog** [comment]
@ZacharyGarrett Thanks! Please close this issue since I'm not be able to log in my own GIthub account recently.
